packages/material-ui/src/ButtonBase/focusVisible.js

```diff
@@ -1,75 +1,111 @@
-import warning from 'warning';
-import ownerDocument from '../utils/ownerDocument';
+// based on https://github.com/WICG/focus-visible/blob/v4.1.5/src/focus-visible.js
 
-const internal = {
-  focusKeyPressed: false,
-  keyUpEventTimeout: -1,
+let hadKeyboardEvent = true;
+let hadFocusVisibleRecently = false;
+let hadFocusVisibleRecentlyTimeout = null;
+
+const inputTypesWhitelist = {
+  text: true,
+  search: true,
+  url: true,
+  tel: true,
+  email: true,
+  password: true,
+  number: true,
+  date: true,
+  month: true,
+  week: true,
+  time: true,
+  datetime: true,
+  'datetime-local': true,
 };
 
-function findActiveElement(doc) {
-  let activeElement = doc.activeElement;
-  while (activeElement && activeElement.shadowRoot && activeElement.shadowRoot.activeElement) {
-    activeElement = activeElement.shadowRoot.activeElement;
+/**
+ * Computes whether the given element should automatically trigger the
+ * `focus-visible` class being added, i.e. whether it should always match
+ * `:focus-visible` when focused.
+ * @param {Element} node
+ * @return {boolean}
+ */
+function focusTriggersKeyboardModality(node) {
```

Comment:
In the specification, https://www.w3.org/TR/selectors-4/#the-focus-visible-pseudo we can read the following:

> The :focus-visible pseudo-class applies while an element matches the :focus pseudo-class and the user agent determines via **heuristics** that the **focus should be made evident** on the element.

I have highlighted the two important parts (heuristics, focus should be made evident). We can later read:

> User agents can choose their own heuristics for when to match :focus-visible; however, the following (non-normative) suggestions can be used as a starting point:

We have an area of freedom we can take advantage of. We can take each point individually.

1. "User preference". it's too early, we can't do anything.
2. "Any element which supports keyboard input". We have no use cases (as far as I know) for using this logic with an element that accepts keyboard input. We can ignore this case, so probably most of this function.
3. "If the user interacts with the page via the keyboard". I believe #15398 support it correctly.
4. "If the user interacts with the page via a pointing device". I believe #15398 support it correctly.
5. "If the active element matches :focus-visible, and a script causes focus to move elsewhere". I believe #15398 support it correctly.
6. "Conversely, if the active element does not match :focus-visible, and a script causes focus to move elsewhere" . I believe #15398 support it correctly.

I like the implementation in #15398. It's short, elegant and seems to cover all the requirements of the specification. I like how this pull request brings most of its logic back to the button without the hook migration requirement.

Reply:
I'd rather trust the W3C implementation of their own spec. Even if they're different people they work under the same umbrella whereas we just stumbled over this issue.

Reply:
@robdodson was already kind enough to help us in the past with the focus logic. Let's try to ask him nicely. Hi Rob, we are looking into improving Material-UI focus visible logic. Do you think that your focus-visible.js could be simplified in the eventuality it hasn't to support "editable" elements? Thanks.

Reply:
Hey @oliviertassinari I'm not sure I follow :)

What use case are you trying to solve?

Reply:
> What use case are you trying to solve?

@robdodson Currently, Material-UI only uses focus-visible logic/styling on various types of buttons and elements that can be interacted with in a button-like manner such as clickable list/menu items. Material-UI doesn't currently use the focus-visible concept for styling on any text inputs or other types of editable text elements, so @oliviertassinari is asking whether there is any harm in dropping the `focusTriggersKeyboardModality` functionality from our version of focus-visible.js.

Reply:
Hey Ryan, if I'm understanding correctly that would result in behavior where clicking on `<input type=text>` would no longer apply the `.focus-visible` class?

The actual native `:focus-visible` selector will match in this scenario so, while you can remove it from the polyfill if you wish, eventually it will be the native behavior.

If y'all can provide an example of where that native behavior would be undesirable I'd really like to hear it. The primary reason we have the polyfill is to get this kind of feedback.

Reply:
@robdodson Rob, I don't think it's a case of being undesirable -- just currently unneeded within Material-UI. In the case of a text input, if I'm understanding correctly, it would **always** receive `.focus-visible` whenever it has `:focus` which means that Material-UI can always do the styling based on `:focus` for those elements. I believe @oliviertassinari is just looking to trim out any pieces we don't need to minimize the bundle size of Material-UI. We only need to leverage `.focus-visible` for element types where it could be out-of-sync with `:focus`, so we just want to ensure that dropping the `focusTriggersKeyboardModality` functionality won't have any adverse effects on those cases.

Reply:
> In the case of a text input, if I'm understanding correctly, it would always receive .focus-visible whenever it has :focus which means that Material-UI can always do the styling based on :focus for those elements.

Yep, correct.

> @oliviertassinari is just looking to trim out any pieces we don't need to minimize the bundle size of Material-UI.

OK cool, understood.

>  we just want to ensure that dropping the focusTriggersKeyboardModality functionality won't have any adverse effects on those cases.

If I recall... I think we always call this function but it's only purpose is to flag those elements where we always want to show a ring. A good heuristic to follow is "if the element would display a keyboard on mobile, then we'll put it in here." We don't have any plans to add additional functionality to that function. Aside from making sure that meta keys don't trigger focus-visible https://github.com/WICG/focus-visible/pull/184 I don't think we have any other big changes planned.

Reply:
> A good heuristic to follow is "if the element would display a keyboard on mobile, then we'll put it in here."

@robdodson Thank you, that's gold!